eclass/rust.eclass

```diff
@@ -0,0 +1,480 @@
+			if [[ "${LLVM_COMPAT[@]}" == *"${llvm_slot}"* ]]; then
+				# We can check for the USE
+				use llvm_slot_${llvm_slot} && llvm_r1_slot=${llvm_slot}
+			else
+				continue
+			fi
+		done
+		if [[ -z ${llvm_r1_slot} ]]; then
+			die "${FUNCNAME}: no LLVM slot found"
+		fi
+	fi
+
+	# iterate over known slots, newest first
+	for slot in "${_RUST_SLOTS_ORDERED[@]}"; do
+		llvm_slot=${_RUST_KNOWN_SLOTS[${slot}]}
+		# skip higher slots
+		if [[ -n ${max_slot} ]]; then
+			if ver_test ${slot} -eq ${max_slot}; then
+				max_slot=
+			elif ver_test ${slot} -gt ${max_slot}; then
+				continue
+			fi
+		fi
+
+		# If we're in LLVM mode we can skip any slots that don't match the selected USE
+		if [[ -n ${RUST_NEEDS_LLVM} ]]; then
+			if [[ ${llvm_slot} != ${llvm_r1_slot} ]]; then
+				continue
+			fi
+		fi
+
+		if declare -f rust_check_deps >/dev/null; then
+			local RUST_SLOT=${slot}
+			local LLVM_SLOT=${_RUST_KNOWN_SLOTS[${slot}]}
+			rust_check_deps && return
+		else
+			local rust_type
+			# Check for an appropriate Rust version and its type.
+			# Prefer the from-source version "because"
+			if (has_version ${hv_switch} "dev-lang/rust:${slot}" ||
+				has_version ${hv_switch} "dev-lang/rust-bin:${slot}"); then
+				if has_version ${hv_switch} "dev-lang/rust:${slot}"; then
+					rust_type="source"
+				else
+					rust_type="binary"
+				fi
+				echo ${slot}
+				echo ${rust_type}
+				return
+			fi
+		fi
+
+		# We want to process the slot before escaping the loop if we've hit the minimum slot
+		if [[ -n ${RUST_MIN_VER} ]]; then
+			if ver_test ${slot} -eq ${RUST_MIN_VER}; then
+				break
+			fi
+		fi
+
+	done
+
+	# max_slot should have been unset in the iteration
+	if [[ -n ${max_slot} ]]; then
+		die "${FUNCNAME}: invalid max_slot=${max_slot}"
+	fi
+
+	die "No Rust slot${1:+ <= ${1}} satisfying the package's dependencies found installed!"
+}
+
+# @FUNCTION: get_rust_prefix
+# @USAGE: [-b|-d]
+# @DESCRIPTION:
+# Find the newest Rust install that is acceptable for the package,
+# and print an absolute path to it. If both -bin and regular Rust
+# are installed, the regular Rust is preferred.
+#
+# The options and behavior are the same as get_rust_slot.
+get_rust_prefix() {
+	debug-print-function ${FUNCNAME} "$@"
+
+	local prefix=${ESYSROOT}
+	[[ ${1} == -b ]] && prefix=${BROOT}
+
+	local slot rust_type
+	{ read -r slot; read -r rust_type; } <<< $(get_rust_slot)
+
+	if [[ ${rust_type} == "source" ]]; then
+		echo "${prefix}/usr/lib/rust/${slot}/"
+	else
+		echo "${prefix}/opt/rust-bin-${slot}/"
+	fi
+}
+
+# @FUNCTION: rust_prepend_path
+# @USAGE: <slot> <type>
+# @DESCRIPTION:
+# Prepend the path to the specified Rust to PATH and re-export it.
+rust_prepend_path() {
+	debug-print-function ${FUNCNAME} "$@"
+
+	[[ ${#} -ne 2 ]] && die "Usage: ${FUNCNAME} <slot> <type>"
+	local slot=${1}
+
+	local rust_path
+	if [[ ${2} == "source" ]]; then
+		rust_path=${ESYSROOT}/usr/lib/rust/${slot}/bin
+	else
+		rust_path=${ESYSROOT}/opt/rust-bin-${slot}/bin
+	fi
```

Comment:
Are you sure that this is valid? That part of LLVM eclasses is a broken mess, taking advantage of stupid CMake behavior. You can't actually expect to be able to run the compilers from `ESYSROOT`.